File: app/core/types/fields/field.py
from abc import ABC, abstractmethod
from typing import TypeVar, Generic
from ..type_enum import FieldType
from ..predicate import Predicate

T = TypeVar('T')


class Field(ABC, Generic[T]):
    """
    Abstract base class for all field types in the database.

    A field represents a single value in a tuple (row). Each field has:
    - A type (int, string, etc.)
    - A value
    - Methods for comparison, serialization, and hashing

    This follows the Template Method pattern - subclasses implement
    specific behavior while the base class defines the interface.
    """

    @abstractmethod
    def get_value(self) -> T:
        """
        Get the value stored in this field.

        Returns:
            The value of the field with its appropriate type
        """
        pass

    @abstractmethod
    def serialize(self) -> bytes:
        """
        Convert this field to bytes for storage on disk.
        """
        pass

    @classmethod
    @abstractmethod
    def deserialize(cls, data: bytes) -> 'Field':
        """
        Create field instance from serialized bytes.

        Args:
            data: The bytes to deserialize

        Returns:
            Field instance

        Raises:
            ValueError: If data is invalid or corrupted
        """
        pass

    @classmethod
    @abstractmethod
    def get_size(cls) -> int:
        """
        Get the fixed size in bytes for this field type.

        Returns:
            Size in bytes
        """
        pass

    @abstractmethod
    def get_type(self) -> FieldType:
        """
        Return the type of this field.
        """
        pass

    @abstractmethod
    def compare(self, predicate: Predicate, other: 'Field') -> bool:
        """
        Compare this field with another using the given predicate.
        """
        pass

    @abstractmethod
    def __str__(self) -> str:
        """
        String representation of the field value.
        """
        pass

    @abstractmethod
    def __repr__(self) -> str:
        """
        Unambiguous string representation for debugging.
        """
        pass

    @abstractmethod
    def __eq__(self, other: object) -> bool:
        """
        Check equality with another field.
        """
        pass

    @abstractmethod
    def __hash__(self) -> int:
        """
        Hash value for this field (needed for sets/dicts).
        """
        pass

File: pkg/types/field.go
package types

import (
	"io"
	"storemy/pkg/primitives"
)

// Field represents a generic field that can be serialized, compared, and hashed.
// It provides methods for type checking, string representation, equality, and comparison operations.
type Field interface {
	// Serialize writes the field's data to the provided writer.
	Serialize(w io.Writer) error

	// Compare evaluates the field against another field using the given primitives.Predicate operation.
	// Returns true if the comparison holds, or an error if the operation is invalid.
	Compare(op primitives.Predicate, other Field) (bool, error)

	// Type returns the type of the field.
	Type() Type

	// String returns a string representation of the field.
	String() string

	// Equals checks if the field is equal to another field.
	Equals(other Field) bool

	// Hash computes a hash value for the field.
	// Returns the hash as a uint32, or an error if hashing fails.
	Hash() (primitives.HashCode, error)

	// Length returns the length of the field's data in bytes.
	// This is useful for serialization and storage purposes.
	Length() uint32
}
